### ISSUE TITLE
AWS ECR credentials not fetched properly

### ISSUE THREAD

**Issue by webratz** [dependabot/dependabot-core#1009]
Hi,

I just tried out the ECR support and found out that you expect hardcoded AWS credentials.
While that is a valid use case if you are running on-prem and need to access AWS resources, it is considered a bad practice when running on AWS.

https://github.com/dependabot/dependabot-core/blob/master/docker/lib/dependabot/docker/utils/credentials_finder.rb#L39

I'd suggest (to stay backwards compatible) that if neither username and password are set (or empty) the default SDK client should be used directly, so it tries to either use ENV variables or the instance role.
WIth that on AWS the need of storing static credentials goes away.

**Comment by greysteil** [comment]
Thanks for the feedback @webratz, but I'm not sure I entirely understand what you're suggesting here. Are you using the library for a location where you can't use ENV variables when building the credentials array that you're passing to it?

**Comment by webratz** [comment]
Hi,

I can use ENV variables, altough using static credentials is not considered a good practice if you are running within an AWS environment (see https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html )
The use of those is restricted / prohibited in several corp environments.
By default the AWS SDKs try to retrieve credentials on their own, if they are not set explicit. (eg from the instance IAM role, or from ENV vars).
I suggest using this way as a fallback to not break the current behaviour, but be more compliant to AWS best practices and security suggestions.

**Comment by greysteil** [comment]
I'm still a bit uncertain on this one.

Our advice for folks using ECR is to create an IAM User with read-only access to ECR and then create access keys for it and give them to Dependabot. This gives Dependabot only the permissions it needs and is long lived (which we require). I think that's the right advice for anyone using dependabot.com.

I'm guessing your use case here is slightly different, and you're talking about a situation where you control the environment that Dependabot is deployed. For that case it sounds like there may already be short-lived credentials present in the environment that Dependabot could tap into.

I'd be happy to accept a PR that extracts those creds from the environment if they're present. `CredentialsFinder` is definitely the place to do it - those creds will then be fed to the Docker registry client we use to access the Docker API and fetch the available tags etc. (note: we *don't* use an ECR client to do that - we use generic HTTP API requests).

**Comment by webratz** [comment]
Indeed, this is only relevant if you run dependabot-core within an AWS environment.
In that case the instance role is taking care of the scoping.
We currently found a workaround with the existing options, but when we need to touch this the next time i'll add a PR

**Comment by greysteil** [comment]
Awesome! Have renamed the issue and will keep it open.

**Comment by JimNero009** [comment]
+1 for this -- if I understand correctly, the current code is _not_ correctly set up to be appropriately used with a task role on something like Amazon ECS and you instead have to give the code an access key pair?

**Comment by thiemok** [comment]
+1 as well, having this would be great.

**Comment by rnorth** [comment]
Sorry to butt in with a tangential suggestion, but: Docker does already have a mechanism for abstracting away the complexity of obtaining registry credentials, in the form of credential helpers. 

I just wonder if delegating all of this to the credential helper/credential store mechanism might make all issues of Docker auth environment discovery disappear, and solve the problem for all kinds of Docker registry, not just ECR...

**Comment by greysteil** [comment]
I'm still 👍on receiving a PR on this one. As discussed above, this is only relevant to folks that are self-hosting Dependabot Core on AWS and don't want to provide credentials in the way Dependabot currently supports.

**Comment by thiemok** [comment]
If anyone wants to pick this up, it should be fairly straight-forward to add by implementing the other credential types specified in
https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/ECR/Client.html#initialize-instance_method
here
https://github.com/dependabot/dependabot-core/blob/b09aea3aaacedc44be3a37e313cd0db86512d7d1/docker/lib/dependabot/docker/utils/credentials_finder.rb#L33
I'm currently lacking a proper setup to test it (and the ruby knowledge to get that setup quickly), but might still look into doing that if I find a free afternoon to get that all setup.

**Comment by JimNero009** [comment]
Hold on stale bot -- I'm gonna have a stab at this.

**Comment by JimNero009** [comment]
Something like: https://github.com/dependabot/dependabot-core/pull/1463

**Comment by afirth** [comment]
PR open for 2 years

**Comment by jeffwidman** [comment]
👋 Maintainer here. 

We've got #6622 open which should resolve this... but I don't have an ECR instance available to test (nor the time to directly test it myself, since this PR is really for the benefit of the community, not GitHub).

Can you folks who are following this issue hoping it'll get fixed give #6622 a spin to confirm it works for you? Leave feedback over there please...

**Comment by vgpastor** [comment]
The problem with @jeffwidman code it's that if we run Dependanbot in Github machines this mahines don't have de awscli logged in.

I propose that we use the AWS roles method to connect in this scenarios. https://docs.aws.amazon.com/es_es/IAM/latest/UserGuide/id_roles.html

For example in Github actions you can use https://github.com/marketplace/actions/aws-assume-role-action to login the instance.

**Comment by jeffwidman** [comment]
The scope of this ticket was limited to running on AWS.

For running within the Dependabot service at GitHub, see:
* https://github.com/dependabot/dependabot-core/issues/5686#issuecomment-1421723135
* https://github.com/dependabot/dependabot-core/issues/6580
